File: algorithms/cpp/ContainerWithMostWater.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int start = 0;
        int end = height.size() - 1;
        int max_area = 0;
        
        while (start < end) {
            int length = end - start;
            int area = min(height[start], height[end]) * length;
            max_area = max(area, max_area);
            if (height[start] < height[end]) start++;
            else end--;
        }
        return max_area;
    }
};

File: algorithms/go/ContainerWithMostWater.go
func maxArea(height []int) int {
	start := 0
	end := len(height) - 1
	max_area := 0

	for start < end {
		length := end - start

		var min_height int
		if height[start] > height[end] {
			min_height = height[end]
		} else {
			min_height = height[start]
		}

		area := min_height * length
		if area > max_area {
			max_area = area
		}
		if height[start] > height[end] {
			end--
		} else {
			start++
		}

	}

	return max_area
}